server/src/main/java/org/elasticsearch/search/aggregations/MultiBucketConsumerService.java

```diff
@@ -38,14 +40,27 @@
  */
 public class MultiBucketConsumerService {
     public static final int DEFAULT_MAX_BUCKETS = 10000;
+    public static final int DEFAULT_CHECK_BUCKETS_STEP_SIZE = 1000;
     public static final Setting<Integer> MAX_BUCKET_SETTING =
         Setting.intSetting("search.max_buckets", DEFAULT_MAX_BUCKETS, 0, Setting.Property.NodeScope, Setting.Property.Dynamic);
 
-    private volatile int maxBucket;
+    public static final Setting<Integer> CHECK_BUCKETS_STEP_SIZE_SETTING =
+        Setting.intSetting("search.check_buckets_step_size", DEFAULT_CHECK_BUCKETS_STEP_SIZE,
+            -1, Setting.Property.NodeScope, Setting.Property.Dynamic);
```

Comment:
I'd be in favor of not making it configurable at all and check every 1000 buckets all the time? (Or maybe 1024 so that the `% 1000` can be replaced with a lighter `& 0x3FF` mask)